src/libraries/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -924,6 +924,10 @@ private static unsafe void EnsureInitializedCore()
 
                     // Mark us as initialized
                     s_initialized = true;
+
+                    // InitializeTerminalAndSignalHandling will reset the terminal on a normal exit.
+                    // This also resets it for termination due to an unhandled exception.
+                    AppDomain.CurrentDomain.UnhandledException += delegate { Interop.Sys.UninitializeTerminal(); };
```

Comment:
I think that we should do that before the `s_initialized` is set to true (so setting `s_initialized` to `true` is the last thing that is part of the initialization logic)

Reply:
We won't start modifying because there is a lock, but in theory once the flag is set to true some other thread can assume everything is initialized and it can be used.

https://github.com/dotnet/runtime/blob/5abf58221b725acc12df0b0c4e225e4d3c3c3b45/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L868-L874

I just like to have all initialization flags set to true when all the initialization logic is finished. I won't block the PR without it, but I think that in the future somebody can add some new initialization logic after your new logic and then we are going to be in trouble if we miss that in a code review.

Reply:
> theory once the flag is set to true some other thread can assume everything is initialized and it can be used.

Does it mean that registering after `s_initialized` is set increases chances that the app will terminate without resetting the terminal?

Reply:
> Does it mean that registering after `s_initialized` is set increases chances that the app will terminate without resetting the terminal?

Based on the logic for `AppDomain.CurrentDomain` I believe it's possible to be so (it seems to be non-trivial and can take some time, so in meantime something can use the console and then somehow crash the app).

https://github.com/dotnet/runtime/blob/5abf58221b725acc12df0b0c4e225e4d3c3c3b45/src/libraries/System.Private.CoreLib/src/System/AppDomain.cs#L31-L43

But the chances are very low.

Reply:
BTW this comment made me realize that simplest "Hello World" app size after trimming may grow with this change. I am still not against it, just something to keep in mind.

Reply:
> This is true for a lot of PRs.

I know, but in case of Hello World there are some benchmarks that some people look at when comparing tech stacks. I just wanted to point it out since IIRC we may be having some automation that tracks such regressions. So then the person looking at it will find my comment and close it as "by design".

Reply:
fyi, I explored what the implementation would need to implement it in a native signal handler. The challenge is that runtime will reset to the default SIGABRT handler (in `PROCAbort` by calling `SEHCleanupSignals`) and a handler that was added later for `Console` that chains back to the runtime handler is not called.

Reply:
@tmds big thanks for explaining why we have not relied on SIGABRT!